### ISSUE TITLE
Microsoft.AspNet.Authentication.Cookies.CookieAuthenticationOptions now has a new AccessDeniedPath

### ISSUE THREAD

**Issue by blowdart** [aspnet/Announcements#40]
Rather than have a single `LoginPath` browsers are redirected to with cookie based authorization fails we have introduced a new `AccessDeniedPath` property on `CookieAuthenticationOptions`. 

If the `LoginPath` property is set and a request generates an http error with a status code **401** the request will be redirected to the path specified.

If the `AccessDeniedPath` property is set and a request generates an http error with a status code **403** the request will be redirected to the path specified.

If the properties are not set then the status codes are returned to the browser.

**Comment by dmccaffery** [comment]
I like it!

**Comment by GuardRex** [comment]
Sweet! :+1: Solves my issue at [How to create a custom AuthorizeAttribute to avoid redirect to Azure login? aspnet/Mvc#2698](https://github.com/aspnet/Mvc/issues/2698)

**Comment by GuardRex** [comment]
Setting `AccessDeniedPath` doesn't stop OpenIdConnect middleware from redirecting to AAD login. Where can I file a request for a similar feature for an `AccessDeniedPath` option for `app.UseOpenIdConnectAuthentication`? ... aspnet/Identity? ... aspnet/Security? ... elsewhere?

**Comment by blowdart** [comment]
It only applies to cookie middleware. If you wanted the same in other middleware you should file a feature request in aspnet/Security.

**Comment by brockallen** [comment]
*trolling* Not `ForbiddenPath`? */trolling*

**Comment by Asshiah** [comment]
Yeah, this should be `ForbiddenPath` and not `AccessDeniedPath` according to the specified http status code handled.

This is in fact quite misleading as we would suspect the `AuthorizeAttribute` to redirect to this page (instead of the login page) when the autorization fails when using its method `Fail` and access to a controller or action is denied based on the current authentication (via cookie middleware). This is not the case, it stills redirect to the login page as when the `AuthorizeAttribute` fails, a 401 status code is returned and this is quite right as this is the right status code to return in this case according to http standards.

**Comment by blowdart** [comment]
Unfortunately we're stuck with AccessDeniedPath for back compat reasons.

As for making AccessDeniedPath for 401 that is arguably wrong. In those cases middleware should be returning a 403, which is why we added the new option.

As I've said elsewhere with the new Policy based authorization approach we would hope the need for creating your own AuthorizationFilterAttribute has gone away. If a policy fails we go the 403 route.

**Comment by Asshiah** [comment]
I think you misunderstood what I was saying because I used the same `AccessDeniedPath` name for the option I proposed and the option you have already offered.

My point was that there shoud be 3 configuration options:
* One `LoginPath` that handles 401 status code in responses that occured because the user is not authenticated with any credentials (wrong username or password).
* One `AccessDeniedPath` that handles 401 status code in responses that occured because the user is not allowed to access the requested action based on its current credentials and the provided policy (he has already logged in with valid username and password but does not have the right credentials according to the policy).
* One `ForbiddenPath` that handles 403 status code in responses that occured because the requested action is not available in the application or cannot be accessed independently of the user credentials. There is no use in this case for the user to try to authenticate with different credentials as the result will always be the same for the requested action.

I understands that there could be back compatibility reasons that do not allow you to use such names for the configuration options I propose, but my point is not really about the naming but more about the middleware behaviour that seems wrong to me based on http standards.

A 403 code should not be issued because the user do not have sufficient credentials according to a policy but because the action requested is unauthorized for any users whatever there credentials are.

The thing about the `AuthorizationFilterAttribute` is irrelevant as this is in my opinion a design flow in the behaviour of the middleware.

IMO, if a policy fail, we should not go to 403.

**Comment by blowdart** [comment]
I disagree with the 401 and Authenticate response I'm afraid. All that tells you is what authentication type is expected, although in the case of forms based auth the middleware will intercept and change it to a redirection.

There was some [added clarity](https://tools.ietf.org/html/rfc7235) to the RFCs

*The 401 (Unauthorized) status code indicates that the request has not been applied because it lacks valid authentication credentials for the target resource. The server generating a 401 response MUST send a WWW-Authenticate header field (Section 4.1) containing at least one challenge applicable to the target resource.*

*A server that receives valid credentials that are not adequate to gain access ought to respond with the 403 (Forbidden) status code.*

We view having an identity but not enough rights as valid, but not adequate. However your middleware could always override this and set its own status code, but I'd hazard a guess that's going to confuse the vast majority of people as it's veering from the common understanding.

This discussion is probably best taken to an issue on the security repo, rather than in announcements, as people are going to get spammed with the issue updates. If you want to do that I'll take the discussion there. For now, to avoid the spam, I'm going to lock the issue here as that's what we've decided to do with new announcements.
